src/geos.h

```diff
@@ -13,13 +13,92 @@
 #pragma GCC diagnostic pop
 
 
+/* Macros to setup GEOS Context and error handlers 
+
+Typical PyGEOS pattern in a function that uses GEOS:
+
+GEOS_INIT;   // or GEOS_INIT_THREADS if you use no CPython calls
+
+// call a GEOS function using the context 'ctx'
+result = SomeGEOSFunc(ctx, ...);
+
+// handle an error state
+if (result == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+
+
+finish:
+  GEOS_FINISH;  //  or GEOS_FINISH_THREADS if you use no CPython calls
+
+*/
+
+// Define the error states
+enum {
+	PGERR_SUCCESS,
+	PGERR_NOT_A_GEOMETRY,
+	PGERR_GEOS_EXCEPTION,
+  PGERR_NO_MALLOC,
+  PGERR_GEOMETRY_TYPE
+};
+
+// Define how the states are handled by CPython
+#define GEOS_HANDLE_ERR\
+    if (last_warning[0] != 0) { PyErr_WarnEx(PyExc_Warning, last_warning, 0); }\
+    switch (errstate) {\
+      case PGERR_SUCCESS:\
+        break;\
+      case PGERR_NOT_A_GEOMETRY:\
+        PyErr_SetString(PyExc_TypeError, "One of the arguments is of incorrect type. Please provide only Geometry objects."); break;\
+      case PGERR_GEOS_EXCEPTION:\
+        PyErr_SetString(geos_exception[0], last_error); break;\
+      case PGERR_NO_MALLOC:\
+        PyErr_SetString(PyExc_MemoryError, "Could not allocate memory"); break;\
+      case PGERR_GEOMETRY_TYPE:\
+        PyErr_SetString(PyExc_TypeError, "One of the Geometry inputs is of incorrect geometry type."); break;\
+      default:\
+        PyErr_Format(PyExc_RuntimeError, "Pygeos ufunc returned with unknown error state code %d.", errstate); break;\
+    }
+
+
+// Define initialization / finalization macros
+#define _GEOS_INIT_DEF\
+    char errstate = PGERR_SUCCESS;\
+    char last_error[1024] = "";\
```

Comment:
Are we at risk of buffer overflows here?  It looks like GEOS errors are getting set here directly without checking their length first and truncating.

Reply:
I think not, I used `snprintf` in `geos_error_handler` and `geos_notice_handler`, which can limit the amount of characters written to the buffer.